runtime-components/non-repudiation-api/src/main/scala/com/daml/nonrepudiation/api/v1/CertificatesEndpoint.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.nonrepudiation.api.v1
+
+import java.io.ByteArrayInputStream
+import java.security.cert.{CertificateFactory, X509Certificate}
+
+import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.{Route, StandardRoute, ValidationRejection}
+import com.daml.nonrepudiation.{CertificateRepository, FingerprintBytes}
+import com.google.common.io.BaseEncoding
+import org.slf4j.{Logger, LoggerFactory}
+import spray.json.{DefaultJsonProtocol, RootJsonFormat}
+
+import scala.util.Try
+
+private[api] final class CertificatesEndpoint private (certificates: CertificateRepository)
+    extends CertificatesEndpoint.JsonProtocol {
+
+  import CertificatesEndpoint._
+
+  private def putCertificate(certificate: X509Certificate): Route =
+    handleExceptions(logAndReport(logger)(UnableToAddTheCertificate)) {
+      val fingerprint = certificates.put(certificate)
+      complete(encode(fingerprint))
+    }
+
+  private def getCertificate(fingerprint: FingerprintBytes): Route =
+    handleExceptions(logAndReport(logger)(UnableToRetrieveTheCertificate)) {
+      certificates.get(fingerprint).fold(reject)(certificate => complete(encode(certificate)))
+    }
+
+  private val route: Route =
+    concat(
+      put {
+        decodeRequest {
+          entity(as[CertificatesEndpoint.Certificate]) {
+            decode(_).fold(rejectBadInput, putCertificate)
+          }
+        }
+      },
+      path(Segment) { fingerprint =>
+        get {
+          decode(fingerprint).fold(rejectBadInput, getCertificate)
+        }
+      },
+    )
+
+}
+
+object CertificatesEndpoint {
+
+  def apply(certificates: CertificateRepository): Route =
+    new CertificatesEndpoint(certificates).route
+
+  private val logger: Logger = LoggerFactory.getLogger(classOf[CertificatesEndpoint])
+
+  private[api] val InvalidCertificateString: String =
+    "Invalid upload, the 'certificate' field must contain a URL-safe base64-encoded X.509 certificate"
```

Comment:
what does url-safe mean here? I thought this is in the post body not a query param.

Reply:
https://tools.ietf.org/html/rfc4648#page-7

Reply:
I assumed this was common knowledge but I can add a link to the RFC if you think that makes sense.

Reply:
(the assumption was towards the user, not towards you 😉)

Reply:
> I thought this is in the post body not a query param.

It would be `/v1/certificate/<fingerprint>`.

Reply:
Whoops, sorry, I misread it. Yes, it's in the body. I just preferred to use a single encoding, does that make sense?

Reply:
Sorry for being unclear. I wasn’t asking what url-safe means, I was asking why we care about it if it’s not in the url and therefore also why we base64 encode it.

I’m not opposed to using base64 encoding here as well but I think you can make a reasonable argument that the ux is slightly nicer for sending the unencoded contents in something like a multi-part form request, e.g., you can pass it via `-F "parametername=@filename"` to curl whereas you have to call `base64` yourself.

How about the following approach for now:
1. Remove the comment on url-safe since it is misleading given that it is not in a url (the rest of the comment can stay)
2. Stick with base64 for now. We’re early access so we could break it if we want to and even if we don’t and we decide the multi-part is sufficiently nicer, keeping both around is relatively cheap.

Reply:
Ok.

Reply:
Mh, that would mean, however, that you cannot be guaranteed to be able to lookup a certificate (`GET /v1/certificate/<fingerprint>`) using the fingerprint returned when uploading (`PUT /v1/certificate`). Are you sure this is fine?

Reply:
After some thinking, I'm not 100% sure about this. It would mean we would have places where we use base64 encoding and other where we use base64 URL-safe encoding (`GET /v1/certificate/<fingerprint>`). Are you sure it makes sense to use two different encodings? In particular, if we decide to use base64 encoding in all cases where a URL is not involved, it would mean we would return a base64 encoded fingerprint when uploading a certificate but then we would require a base64 URL-safe encoding to retrieve it.

Reply:
I think I’m confused, I thought this was only about the format used when uploading the certificate. I fully agree that the fingerprint returned in the response to that should be the same format as the one in the URL.

Reply:
Sure, I was wondering if it makes sense to use two different encodings, even though it's not strictly needed to be URL-safe for certificates. Since we're forced to use URL-safe encoding for fingerprints, does it make sense to use a different encoding for certificates? I'm mostly trying to understand what an API user would think would make more sense.

Reply:
I guess I view the certificate as binary data which I already have (the certificate lies around somewhere locally) while I view the fingerprint as an opaque string that is returned and produced by the API. So I think it’s not entirely unreasonable to have different encodings for them. That said, I think two different base64 encodings are confusing as you mentioned and doesn’t solve anything So I think we have the following two options:

1. Use url-safe base64 encoding everywhere, i.e., do what you have. Probably just needs a comment somewhere on why we do this for cases where something is not in a url.
2. Use url-safe base64 encoding for things that are strings and/or used in urls, e.g., fingerprints. Upload binary data like certificates without base64 encoding them first, e.g., as multipart form requests.

I would suggest to go with 1 for now and keep the option open to later add/switch to 2 but I don’t feel strongly about this so if you disagree, happy to go with your approach instead.

Reply:
Ok, I'll go with number 1 for now then and keep URL-safe base64 encoding everywhere. At this point changing this behavior at a later point is really not difficult. Thanks for the feedback!

Reply:
Addressed by https://github.com/digital-asset/daml/pull/8982/commits/e7ef80b57672191ea6bc907dc5032d0d6b39271c